Add support for non-prelude types

Currently, custom types are not working within the DFG, because the generated type casts lack the necessary imports to access the data types:

```
error[E0412]: cannot find type `House` in this scope
  --> src/ohua_runtime/generictype.rs:22:15
   |
22 | impl From<Box<House>> for Box<GenericType> {
   |               ^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
   |
2  | use house::House;
   |
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -40,5 +40,4 @@ script:
   - cd testcases/string_manipulation/ && cargo run && cd ../../
   - cd testcases/tuple_values/ && cargo run && cd ../../
   - cd testcases/double_use/ && cargo run && cd ../../
-
-#   - cd testcases/custom_types/ && cargo run && cd ../../
+  - cd testcases/custom_types/ && cargo run && cd ../../

--- a/src/codegen/mod.rs
+++ b/src/codegen/mod.rs
@@ -35,9 +35,7 @@ pub fn generate_ohua_runtime(
         return Err(CodeGenerationError::TargetDirNotCreated(err));
     }
 
-    // ====== start the real work ======
-
-    // read the data structures
+    // read the DFG data structure
     let dfg_file = File::open(&algorithm.ohuao).unwrap();
     let ohua_data: OhuaData = serde_json::from_reader(dfg_file).unwrap();
 
@@ -64,8 +62,6 @@ pub fn generate_ohua_runtime(
         }
     }
 
-    // TODO: Continue to rewrite into error type, write first `build.rs`, verify it works
-
     // populate the module with the static files
     if let Err(err) = populate_static_files(output.clone()) {
         return Err(CodeGenerationError::StaticPopulationFailed(err));
@@ -83,7 +79,7 @@ pub fn generate_ohua_runtime(
     // generate wrapper functions for all operators
     let altered_ohuadata = match wrappers::generate_wrappers(
         ohua_data,
-        &typeinfo.algo_io.argument_types,
+        &typeinfo,
         (output.clone() + "/wrappers.rs").as_str(),
     ) {
         Ok(data) => data,
@@ -94,7 +90,7 @@ pub fn generate_ohua_runtime(
 
     // generate the module file
     if let Err(err) =
-        modgen::generate_modfile(&typeinfo.algo_io, (output.clone() + "/mod.rs").as_str())
+        modgen::generate_modfile(&typeinfo, (output.clone() + "/mod.rs").as_str())
     {
         return Err(CodeGenerationError::ModfileGenerationFailed(err));
     }

--- a/src/codegen/modgen.rs
+++ b/src/codegen/modgen.rs
@@ -1,5 +1,5 @@
 //! Module File generator
-use ohua_types::AlgorithmArguments;
+use type_extract::TypeKnowledgeBase;
 use std::io::{Result, Write};
 use std::fs::File;
 
@@ -8,22 +8,29 @@ use std::fs::File;
 /// By design, the module file only contains templating placeholders in the entry function header
 /// and the return point, where the provided type signatures of the algorithm are inserted.
 /// Only produces an IO error when unable to open and write to a file.
-pub fn generate_modfile(mainarg_types: &AlgorithmArguments, target_path: &str) -> Result<()> {
+pub fn generate_modfile(ty_knowledge: &TypeKnowledgeBase, target_path: &str) -> Result<()> {
     let mod_file = include_str!("templates/mod.rs");
     let input_send = include_str!("templates/snippets/input_send.in");
+
+    // check if any arguments to the function require an import
+    let mut imports = String::new();
+    for imp in ty_knowledge.find_imports_for_algo_io() {
+        imports += &format!("use {};\n", imp);
+    }
 
     // simultaneously construct the string that contains the function header args and
     // the initial argument handovers to the ohua runtime
     let mut function_arguments = String::new();
     let mut inputs = String::new();
-    for (ind, typename) in mainarg_types.argument_types.iter().enumerate() {
+    for (ind, typename) in ty_knowledge.algo_io.argument_types.iter().enumerate() {
         function_arguments +=
             format!("arg{n}: {typename}, ", n = ind, typename = typename).as_str();
         inputs += input_send.replace("{n}", &ind.to_string()).as_str();
     }
 
-    let mut module = String::from(mod_file).replace("{input_args}", function_arguments.as_str());
-    module = module.replace("{return_type}", mainarg_types.return_type.as_str());
+    let mut module = String::from(mod_file).replace("{ty_imports}", imports.as_str());
+    module = module.replace("{input_args}", function_arguments.as_str());
+    module = module.replace("{return_type}", ty_knowledge.algo_io.return_type.as_str());
     module = module.replace("{send_input}", inputs.as_str());
 
     File::create(target_path)?.write_fmt(format_args!("{}", module))

--- a/src/codegen/templates/generictype.rs
+++ b/src/codegen/templates/generictype.rs
@@ -1,2 +1,3 @@
+{imports}
 #[derive(Debug)]
 pub struct GenericType {}

--- a/src/codegen/templates/mod.rs
+++ b/src/codegen/templates/mod.rs
@@ -11,6 +11,8 @@ use self::types::{Arc, ArcIdentifier, OhuaOperator, ValueType};
 use std::thread::{self, Builder};
 use std::sync::mpsc;
 
+{ty_imports}
+
 
 fn sorted_recv_insertion(recvs: &mut Vec<(u32, mpsc::Receiver<Box<GenericType>>)>, recv: mpsc::Receiver<Box<GenericType>>, target: u32) {
     let mut index: usize = 0;
@@ -152,7 +154,7 @@ pub fn ohua_main({input_args}) -> {return_type} {
                         args.push(content);
                     } else {
                         #[cold]
-                        // when we sre in `exiting` state, we should not be here...
+                        // when we are in `exiting` state, we should not be here...
                         eprintln!("[Error] Thread {} entered an inconsistent state. Some input Arcs are empty, others not.", thread::current().name().unwrap());
                         break 'threadloop;
                     }

--- a/src/codegen/typecasts.rs
+++ b/src/codegen/typecasts.rs
@@ -9,10 +9,22 @@ use std::io::{Result, Write};
 /// Generates a bidirectional type cast for every type from the given hash set.
 ///
 /// Returns an IO Error when opening or writing to a file fails.
-fn generate_for(types: HashSet<String>, target_file: &str) -> Result<()> {
-    let generic_type_file = include_str!("templates/generictype.rs");
+fn generate_for(
+    types: HashSet<String>,
+    import_paths: HashSet<String>,
+    target_file: &str,
+) -> Result<()> {
+    let generic_type_file = String::from(include_str!("templates/generictype.rs"));
     let typecast = include_str!("templates/snippets/typecast.in");
 
+    // Add the imports to the file
+    let mut imports = String::new();
+    for imp in import_paths {
+        imports += &format!("use {};\n", imp);
+    }
+    let type_header = generic_type_file.replace("{imports}", imports.as_str());
+
+    // generate the casts
     let mut typecast_file = String::new();
 
     for datatype in types {
@@ -23,7 +35,8 @@ fn generate_for(types: HashSet<String>, target_file: &str) -> Result<()> {
         );
     }
 
-    File::create(target_file)?.write_fmt(format_args!("{}{}", generic_type_file, typecast_file))
+    // write the imports and the casts concatenated to a file
+    File::create(target_file)?.write_fmt(format_args!("{}{}", type_header, typecast_file))
 }
 
 /// Retrieves the argument types for a given function from the type knowledge base
@@ -57,7 +70,9 @@ pub fn generate_casts(
     typeinfo: &TypeKnowledgeBase,
     target_file: &str,
 ) -> Result<()> {
+    // TODO: Couldn't this be rewritten by omitting the operator iteration and directly using all types from the knowledge base?
     let mut used_types: HashSet<String> = HashSet::new();
+    let mut necessary_imports: HashSet<String> = HashSet::new();
 
     // also make use of the argument types provided from the `type_dump` file
     for arg in &typeinfo.algo_io.argument_types {
@@ -72,5 +87,22 @@ pub fn generate_casts(
         }
     }
 
-    generate_for(used_types, target_file)
+    // gather all necessary import paths
+    for func in &typeinfo.functions {
+        for arg in &func.arguments {
+            if let Some(ref paths) = arg.path {
+                for path in paths {
+                    necessary_imports.insert(path.clone());
+                }
+            }
+        }
+
+        if let Some(ref paths) = func.return_val.path {
+            for path in paths {
+                necessary_imports.insert(path.clone());
+            }
+        }
+    }
+
+    generate_for(used_types, necessary_imports, target_file)
 }

--- a/src/codegen/wrappers.rs
+++ b/src/codegen/wrappers.rs
@@ -5,6 +5,7 @@ use std::io::{Result, Write};
 use std::fs::File;
 
 use ohua_types::*;
+use type_extract::TypeKnowledgeBase;
 
 /// Generates a wrapper for a single stateful function.
 ///
@@ -78,9 +79,7 @@ fn wrap_function(name: &str, incoming_arcs: usize, mut outgoing_arcs: Vec<usize>
                     ).as_ref(),
                 );
             } else {
-                outgoing.push_str(
-                    format!("vec![{}]", "Box::from(Box::new(res))").as_ref(),
-                );
+                outgoing.push_str(format!("vec![{}]", "Box::from(Box::new(res))").as_ref());
             }
         }
 
@@ -102,7 +101,10 @@ fn wrap_function(name: &str, incoming_arcs: usize, mut outgoing_arcs: Vec<usize>
 /// The hashset of namespaces is used to generate the correct imports to have all functions in scope.
 fn analyze_dfg(
     ohuadata: &OhuaData,
-) -> (HashMap<String, (usize, Vec<usize>, Vec<i32>)>, HashSet<String>) {
+) -> (
+    HashMap<String, (usize, Vec<usize>, Vec<i32>)>,
+    HashSet<String>,
+) {
     // the function map tracks the number of I/O ports a function has as well as the operators associated with a function
     // NOTE: there might be more than one operator per function
     let mut function_map: HashMap<String, (usize, Vec<usize>, Vec<i32>)> = HashMap::new();
@@ -184,7 +186,11 @@ fn analyze_dfg(
 /// cloning and a clean integration into the DFG.
 ///
 /// Also creates the necessary new operators for the mainarg wrappers.
-fn generate_mainarg_wrappers(first_id: i32, ohuadata: &OhuaData, mainarg_types: &Vec<String>) -> (Vec<Operator>, String) {
+fn generate_mainarg_wrappers(
+    first_id: i32,
+    ohuadata: &OhuaData,
+    mainarg_types: &Vec<String>,
+) -> (Vec<Operator>, String) {
     let template = include_str!("templates/snippets/mainarg.in");
 
     let mut operators = Vec::new();
@@ -211,12 +217,16 @@ fn generate_mainarg_wrappers(first_id: i32, ohuadata: &OhuaData, mainarg_types: 
             1 => {
                 wrapper = wrapper.replace(
                     "{argument}",
-                    format!("Box::from(Box::new({}))", "arg").as_str())
+                    format!("Box::from(Box::new({}))", "arg").as_str(),
+                )
             }
             _ => {
                 wrapper = wrapper.replace(
                     "{argument}",
-                    format!("Box::from(Box::new({}.clone())), ", "arg").repeat(num_uses).as_str())
+                    format!("Box::from(Box::new({}.clone())), ", "arg")
+                        .repeat(num_uses)
+                        .as_str(),
+                )
             }
         }
         wrapper_code.push_str(wrapper.as_str());
@@ -243,14 +253,18 @@ fn generate_mainarg_wrappers(first_id: i32, ohuadata: &OhuaData, mainarg_types: 
 /// links to the corresponding wrapped functions and add the main argument wrappers.
 ///
 /// Returns either an IO error when opening/writing to the file failed or the updated ohua data structure
-pub fn generate_wrappers(mut ohuadata: OhuaData, mainarg_types: &Vec<String>, target_file: &str) -> Result<OhuaData> {
+pub fn generate_wrappers(
+    mut ohuadata: OhuaData,
+    lookupinfo: &TypeKnowledgeBase,
+    target_file: &str,
+) -> Result<OhuaData> {
     // analyze the dataflow graph
     let (function_map, namespaces) = analyze_dfg(&ohuadata);
 
     let skeleton = include_str!("templates/wrappers.rs");
 
     // generate imports
-    let imports = namespaces
+    let mut imports = namespaces
         .iter()
         .fold(String::new(), |acc, ref x| acc + "use " + x + ";\n");
 
@@ -273,15 +287,26 @@ pub fn generate_wrappers(mut ohuadata: OhuaData, mainarg_types: &Vec<String>, ta
 
     // wrap the main arguments
     let first_mainarg = (ohuadata.graph.operators.len() + 1) as i32;
-    let (mut mainarg_ops, wrapper_code) = generate_mainarg_wrappers(first_mainarg, &ohuadata, mainarg_types);
+    let (mut mainarg_ops, wrapper_code) =
+        generate_mainarg_wrappers(first_mainarg, &ohuadata, &lookupinfo.algo_io.argument_types);
 
     // register each mainarg operator as target for an input arc
     ohuadata.graph.input_targets.reserve(mainarg_ops.len());
     for op in &mainarg_ops {
-        ohuadata.graph.input_targets.push(ArcIdentifier {operator: op.operatorId, index: 0})
+        ohuadata.graph.input_targets.push(ArcIdentifier {
+            operator: op.operatorId,
+            index: 0,
+        })
     }
 
     ohuadata.graph.operators.append(&mut mainarg_ops);
+
+    // after the mainargs have been appended, add the imports for their argument types
+    imports += lookupinfo
+        .find_imports_for_algo_io()
+        .iter()
+        .fold(String::from("\n"), |acc, ref x| acc + "use " + x + ";\n")
+        .as_str();
 
     // rewrite the env arcs for any main arguments encountered
     for mut arc in ohuadata.graph.arcs.iter_mut() {

--- a/src/type_extract/algo_io_imports.rs
+++ b/src/type_extract/algo_io_imports.rs
@@ -1,0 +1,23 @@
+use syn;
+
+use super::parser::{collect_type, FunctionInfo};
+
+/// Given a type, look through the `FunctionInfo` vector and extract any imports that have to be made explicitly
+pub fn find_paths_for_used_types(ty: &syn::Type, infos: &Vec<FunctionInfo>) -> Vec<String> {
+    let wanted_ty = collect_type(ty);
+
+    for info in infos {
+        for arg in &info.arguments {
+            if arg.name == wanted_ty {
+                return arg.path.clone().unwrap_or(vec![]);
+            }
+        }
+
+        if info.return_val.name == wanted_ty {
+            return info.return_val.path.clone().unwrap_or(vec![]);
+        }
+    }
+
+    // TODO: if nothing was found, should we error or pretend it's alright?
+    panic!("No type info for type {} was found.", wanted_ty);
+}

--- a/src/type_extract/mod.rs
+++ b/src/type_extract/mod.rs
@@ -7,6 +7,7 @@ use syn;
 use serde_json;
 
 mod parser;
+mod algo_io_imports;
 
 pub use self::parser::{FunctionInfo, Type};
 use ohuac::OhuaProduction;
@@ -58,6 +59,22 @@ impl TypeKnowledgeBase {
         }
 
         Ok(knowledgebase)
+    }
+
+    /// Collects import paths for all types mentioned in the `algo_io` member
+    pub fn find_imports_for_algo_io(&self) -> Vec<String> {
+        let mut imports = Vec::new();
+
+        for input in &self.algo_io.argument_types {
+            let parsed: syn::Type = match syn::parse_str(input) {
+                Ok(tokens) => tokens,
+                Err(e) => panic!("[Stage 4] A critical error occured while parsing the ohuac type dump output.\n This is an internal error. \n{}", e),
+            };
+
+            imports.append(&mut algo_io_imports::find_paths_for_used_types(&parsed, &self.functions));
+        }
+
+        imports
     }
 }
 

--- a/src/type_extract/parser.rs
+++ b/src/type_extract/parser.rs
@@ -4,6 +4,8 @@ use errors::TypeExtractionError;
 use syn;
 use quote::{ToTokens, Tokens};
 
+// TODO: Clean up this mess
+
 /// Type that contains type lookup information for a module. `K` is the outside name from `V`, when existing.
 type LookupTable = HashMap<String, LookupInfo>;
 
@@ -272,6 +274,56 @@ fn collect_type(argument: &syn::Type) -> String {
     tokens.to_string().replace(" ", "")
 }
 
+/// Inspects a generic argument from a path segment to extract any necessary imports
+fn inspect_angle_bracketed_generic_arg(
+    generic_arg: &syn::AngleBracketedGenericArguments,
+    lookup_table: &LookupTable,
+) -> Vec<String> {
+    let mut imports = Vec::new();
+
+    for arg in &generic_arg.args {
+        // We only care about generic arguments of type "Type" here
+        if let syn::GenericArgument::Type(ref ty) = *arg {
+            imports.append(&mut find_paths_for_type(ty, lookup_table));
+        }
+    }
+
+    imports
+}
+
+/// Inspects a type path to collect all necessary imports for types used within
+fn inspect_path_segments(ty_path: &syn::Path, lookup_table: &LookupTable) -> Vec<String> {
+    let mut imports = Vec::new();
+
+    // check whether the first part of the imported path is found in the DB
+    if let Some(val) =
+        lookup_table.get(&ty_path.segments.first().unwrap().value().ident.to_string())
+    {
+        let mut import = val.import_path.clone();
+        if val.is_renamed() {
+            import += format!(" as {}", val.outside_name.clone().unwrap()).as_str();
+        }
+        imports.push(import);
+    }
+
+    // inspect all path segments to look for path arguments (like in Vec<Type> or Result<Foo, Bar>)
+    for segment in &ty_path.segments {
+        match segment.arguments {
+            syn::PathArguments::None => (),
+            syn::PathArguments::Parenthesized(_) => println!(
+                "[DEBUG] Ignoring parenthesized path arguments at {:?}",
+                segment
+            ),
+            syn::PathArguments::AngleBracketed(ref ang) => {
+                imports.append(&mut inspect_angle_bracketed_generic_arg(ang, lookup_table))
+            }
+        }
+    }
+
+    imports
+}
+
+// TODO: That is a incredibly misleading name
 /// Using a previously generated lookup table, try to collect all necessary imports for a (possibly composited) type
 fn find_paths_for_type(ty: &syn::Type, lookup_table: &LookupTable) -> Vec<String> {
     let mut imports = Vec::new();
@@ -298,22 +350,7 @@ fn find_paths_for_type(ty: &syn::Type, lookup_table: &LookupTable) -> Vec<String
             imports.append(&mut find_paths_for_type(ty, lookup_table));
         },
         Path(ref ty_path) => {
-            // check whether the first part of the imported path is found in the DB
-            if let Some(val) = lookup_table.get(&ty_path
-                .path
-                .segments
-                .first()
-                .unwrap()
-                .value()
-                .ident
-                .to_string())
-            {
-                let mut import = val.import_path.clone();
-                if val.is_renamed() {
-                    import += format!(" as {}", val.outside_name.clone().unwrap()).as_str();
-                }
-                imports.push(import);
-            }
+            imports.append(&mut inspect_path_segments(&ty_path.path, lookup_table))
         }
         TraitObject(_) => {
             panic!("Function arguments bound by Traits are not supported by ohua yet.")

--- a/src/type_extract/parser.rs
+++ b/src/type_extract/parser.rs
@@ -267,7 +267,7 @@ fn resolve_use_stmt(item_use: &syn::UseTree, current_state: Option<LookupInfo>) 
     table
 }
 
-fn collect_type(argument: &syn::Type) -> String {
+pub fn collect_type(argument: &syn::Type) -> String {
     let mut tokens = Tokens::new();
     argument.to_tokens(&mut tokens);
 
@@ -325,7 +325,7 @@ fn inspect_path_segments(ty_path: &syn::Path, lookup_table: &LookupTable) -> Vec
 
 // TODO: That is a incredibly misleading name
 /// Using a previously generated lookup table, try to collect all necessary imports for a (possibly composited) type
-fn find_paths_for_type(ty: &syn::Type, lookup_table: &LookupTable) -> Vec<String> {
+pub fn find_paths_for_type(ty: &syn::Type, lookup_table: &LookupTable) -> Vec<String> {
     let mut imports = Vec::new();
 
     use syn::Type::*;

--- a/testcases/custom_types/custom_types.ohuac
+++ b/testcases/custom_types/custom_types.ohuac
@@ -1,14 +1,17 @@
 ns some_ns;
 
-use sf house (move_house, move_in_one, evict_one, house_information);
+use sf house (move_house, move_in_one, evict_one, house_information, check_any_humans_left);
 
 fn main(house: House, target_address: String, humans: Vec<Human>) -> House {
-    let house = move_house(house, target_address);
-    let (house, humans) = move_in_one(house, humans);
-    let (house, humans) = move_in_one(house, humans);
-    let (house, humans) = move_in_one(house, humans);
+    let house2 = move_house(house, target_address);
+    let (house3, humans2) = move_in_one(house2, humans);
+    let (house4, humans3) = move_in_one(house3, humans2);
+    let (house5, humans4) = move_in_one(house4, humans3);
 
-    house_information(house);
+    house_information(house5);
 
-    evict_one(house)
+    // TODO: remove this workaround when issue with unused operator outputs is resolved!
+    check_any_humans_left(humans4);
+
+    evict_one(house5)
 }

--- a/testcases/custom_types/src/house.rs
+++ b/testcases/custom_types/src/house.rs
@@ -7,7 +7,7 @@ pub struct Human {
 pub struct House {
     pub rooms: u8,
     pub inhabitants: Vec<Human>,
-    address: String
+    pub address: String
 }
 
 pub enum EmotionalState {
@@ -17,19 +17,19 @@ pub enum EmotionalState {
 }
 
 impl House {
-    pub fn getAddress(&self) -> String {
+    pub fn get_address(&self) -> String {
         self.address.clone()
     }
 
-    pub fn changeAddress(&mut self, new: String) {
+    pub fn change_address(&mut self, new: String) {
         self.address = new;
     }
 }
 
 pub fn move_house(mut house: House, target_address: String) -> House {
-    print!("Moving the House from {}. ", house.getAddress());
-    house.changeAddress(target_address);
-    println!("Moved to {}", house.getAddress());
+    print!("Moving the House from {}. ", house.get_address());
+    house.change_address(target_address);
+    println!("Moved to {}", house.get_address());
     house
 }
 
@@ -41,7 +41,7 @@ pub fn move_in_one(mut house: House, mut humans: Vec<Human>) -> (House, Vec<Huma
 }
 
 pub fn house_information(house: House) {
-    println!("Address: {}, Rooms: {}", house.getAddress(), house.rooms);
+    println!("Address: {}, Rooms: {}", house.get_address(), house.rooms);
     let mut inhabitants = String::new();
     for inh in house.inhabitants {
         inhabitants += (inh.name + ", ").as_str();
@@ -52,4 +52,8 @@ pub fn house_information(house: House) {
 pub fn evict_one(mut house: House) -> House {
     let _ = house.inhabitants.pop();
     house
+}
+
+pub fn check_any_humans_left(_humans: Vec<Human>) {
+    ();
 }

